rust/template/differential_datalog/src/program/mod.rs

```diff
@@ -85,7 +89,111 @@ type TKeyAgent<S> = TraceAgent<KeyTrace<S>>;
 type TValEnter<P, T> = TraceEnter<TValAgent<P>, T>;
 type TKeyEnter<P, T> = TraceEnter<TKeyAgent<P>, T>;
 
-/// Diff associated with records in differential dataflow
+#[derive(
+    Abomonation, Copy, Ord, PartialOrd, Eq, PartialEq, Hash, Debug, Clone, Serialize, Deserialize,
+)]
+pub struct CheckedWeight {
+    pub value: i32,
+}
+
+/// CheckedWeight trait implementations
+
+impl Semigroup for CheckedWeight {
+    fn is_zero(&self) -> bool {
+        self.value == 0
+    }
+}
+
+impl<'a> AddAssign<&'a Self> for CheckedWeight {
+    fn add_assign(&mut self, other: &'a Self) {
+        // intentional panic on overflow
+        self.value = self
+            .value
+            .checked_add(other.value)
+            .expect("Weight overflow");
+    }
+}
+
+impl Add for CheckedWeight {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        // intentional panic on overflow
+        CheckedWeight {
+            value: self
+                .value
+                .checked_add(other.value)
+                .expect("Weight overflow"),
+        }
+    }
+}
+
+impl Mul for CheckedWeight {
+    type Output = Self;
+    fn mul(self, rhs: Self) -> Self::Output {
+        // intentional panic on overflow
+        CheckedWeight {
+            value: self.value.checked_mul(rhs.value).expect("Weight overflow"),
+        }
+    }
+}
+
+impl Neg for CheckedWeight {
+    type Output = Self;
+
+    fn neg(self) -> Self::Output {
+        // intentional panic on overflow
+        CheckedWeight {
+            value: 0_i32.checked_sub(self.value).expect("Weight overflow"),
+        }
+    }
+}
+
+impl Monoid for CheckedWeight {
+    fn zero() -> Self {
+        CheckedWeight { value: 0 }
+    }
+}
+
+impl One for CheckedWeight {
+    fn one() -> Self {
+        CheckedWeight { value: 1 }
+    }
+}
+
+impl Zero for CheckedWeight {
+    fn zero() -> Self {
+        CheckedWeight { value: 0 }
+    }
+    fn is_zero(&self) -> bool {
```

Comment:
```suggestion
        Self { value: 0 }
    }

    fn is_zero(&self) -> bool {
```

Reply:
Do you want to apply @Kixiron 's suggestion here and in other places below?